engine/src/btc/cached_rpc.rs

```diff
@@ -0,0 +1,153 @@
+use crate::{
+	btc::{
+		retry_rpc::BtcRetryRpcClient,
+		rpc::{BlockHeader, BtcRpcApi, VerboseBlock},
+	},
+	caching_request::CachingRequest,
+};
+use bitcoin::{BlockHash, Txid};
+use cf_chains::btc::{BlockNumber, BtcAmount};
+use cf_utilities::task_scope::Scope;
+use tokio::sync::mpsc;
+
+#[derive(Clone)]
+pub struct BtcCachingClient {
+	block: CachingRequest<BlockHash, VerboseBlock, BtcRetryRpcClient>,
+	block_hash: CachingRequest<BlockNumber, BlockHash, BtcRetryRpcClient>,
+	send_raw_transaction: CachingRequest<Vec<u8>, Txid, BtcRetryRpcClient>,
+	next_block_fee: CachingRequest<(), Option<BtcAmount>, BtcRetryRpcClient>,
+	avg_fee_rate: CachingRequest<BlockHash, BtcAmount, BtcRetryRpcClient>,
+	block_header: CachingRequest<BlockHash, BlockHeader, BtcRetryRpcClient>,
+	best_block_hash: CachingRequest<(), BlockHash, BtcRetryRpcClient>,
+
+	pub cache_invalidation_senders: Vec<mpsc::Sender<()>>,
+}
+
+impl BtcCachingClient {
+	pub(crate) fn new(scope: &Scope<'_, anyhow::Error>, client: BtcRetryRpcClient) -> Self {
+		let (block, block_cache) =
+			CachingRequest::<BlockHash, VerboseBlock, BtcRetryRpcClient>::new(
+				scope,
+				client.clone(),
+			);
+		let (block_hash, block_hash_cache) =
+			CachingRequest::<BlockNumber, BlockHash, BtcRetryRpcClient>::new(scope, client.clone());
+		let (send_raw_transaction, send_raw_transaction_cache) =
+			CachingRequest::<Vec<u8>, Txid, BtcRetryRpcClient>::new(scope, client.clone());
+		let (next_block_fee, next_block_fee_cache) =
+			CachingRequest::<(), Option<BtcAmount>, BtcRetryRpcClient>::new(scope, client.clone());
```

Comment:
This seems like a request that we might not want to cache - unless the cache is cleared regularly?

Reply:
As discussed, we clear the cache on every SC block before we start computing our votes. See usage of `self.cache_invalidation_senders` in `engine/elections`.